### PR TITLE
[Part2 of #4895] Changed ComplexTyepModelBinderProvider to check for parameterless constructor

### PR DESCRIPTION
**Pull request by kichalla**
@dougbu @rynowak

### CODE REVIEW

File: test/Microsoft.AspNetCore.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 1]
```diff
@@ -383,6 +383,68 @@ private class Person5
...
+            var testContext = ModelBindingTestHelper.GetTestContext(request =>
+            {
+                request.QueryString = QueryString.Create(key, "James");
+            });
+
+            // Act & Assert
+            var exception = await Assert.ThrowsAsync<InvalidOperationException>(async () =>
+            {
+                await argumentBinder.BindModelAsync(parameter, testContext);
+            });
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: Get rid of the ceremonial braces and `async` / `await`: `await Assert.ThrowsAsync<...>(() => argumentBinder.BindModelAsync(parameter, testContext));`

File: test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ComplexTypeModelBinderProviderTest.cs [hunk 2]
```diff
@@ -55,11 +55,81 @@ public void Create_ForSupportedTypes_ReturnsBinder()
...
+            var context = new TestModelBinderProviderContext(typeof(AbstractClassWithDefaultConstructor));
+
+            // Act
+            var result = provider.GetBinder(context);
+
+            // Assert
+            Assert.Null(result);
+        }
+
+        // Actually c# doesn't allow creating parameterless constructors for structs.
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Binders/ComplexTypeModelBinderProviderTest.cs]: The name of this `struct` is incorrect. Actually C# _always_ defines a parameterless constructor for `struct`s and developers can't override that one. Suggest debugging to see exactly how `ModelMetadata` describes this type to the `ComplexTypeModelBinderProvider`.

File: test/Microsoft.AspNetCore.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 3]
```diff
@@ -383,6 +383,68 @@ private class Person5
...
+
+            Assert.Equal(
+                string.Format("Could not create a model binder for model object of type '{0}'.", parameterType.FullName),
+                exception.Message);
+        }
+
+        private struct StructWithNoDefaultConstructor
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: Again, come up with a more correct name.

File: test/Microsoft.AspNetCore.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 4]
```diff
@@ -383,6 +383,68 @@ private class Person5
             Assert.Empty(modelState.Keys);
         }
 
+        [Theory]
+        [InlineData(typeof(ClassWithNoDefaultConstructor), "Name")]
+        [InlineData(typeof(StructWithNoDefaultConstructor), "Name")]
+        [InlineData(typeof(AbstractClassWithNoDefaultConstructor), "Name")]
+        public async Task ActionParameter_NoDefaultConstructor_Fails(Type parameterType, string key)
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: Why is the `key` an argument?

**Comment by kichalla** [comment]
🆙 📅

**Comment by kichalla** [comment]
🆙 📅

File: test/Microsoft.AspNetCore.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 5]
```diff
@@ -383,6 +383,120 @@ private class Person5
...
+
+            // Assert
+            Assert.True(result.IsModelSet);
+            Assert.True(modelState.IsValid);
+            var model = Assert.IsType<Class1>(result.Model);
+            Assert.Null(model.Property1);
+            var key = Assert.Single(modelState.Keys);
+            Assert.Equal("Name", key);
+            Assert.Equal("James", modelState[key].AttemptedValue);
+            Assert.Equal("James", modelState[key].RawValue);
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: This is a bizarre pattern. If you copied this from another test, great if you could fix that too.
``` c#
var keyValuePair = Assert.Single(modelState);
Assert.Equal("Name", keyValuePair.Key);
Assert.Equal("James", keyValuePair.Value.AttemptedValue);
Assert.Equal("James", keyValuePair.Value.RawValue);
```

Much less important, good to specify `StringComparer.Ordinal` even in test assertions (when not ignoring case of course).

**Comment by kichalla** [comment]
d09e921c4a896865b0bd81bd99e8e69dc19460b7

**Pull request closed without merging**